persistent_queue/persistent_queue.py

```diff
@@ -0,0 +1,464 @@
+
+    def put_nowait(self, item):
+        """
+        Provides compatibility with stdlib Queue objects.
+
+        Equivalent to put(item, False).
+        """
+        self.put(item, block=False)
+
+    def get(self, block=True, timeout=None, items=1):
+        """
+        Provides compatibility with stdlib Queue objects.
+        items: number of how many items are returned. If items is greater than
+        one, a list is returned.
+
+        Remove and return an item from the queue. If optional args block is
+        true and timeout is None (the default), block if necessary until an
+        item is available. If timeout is a positive number, it blocks at most
+        timeout seconds and raises the Empty exception if no item was available
+        within that time. Otherwise (block is false), return an item if one is
+        immediately available, else raise the Empty exception (timeout is
+        ignored in that case).
+        """
+        _LOGGER.debug("Getting %s items", items)
+
+        # Ignore requests for zero items
+        if items == 0:
+            _LOGGER.debug("Returning empty list")
+            return []
+
+        with self._get_lock:
+            data, queue_top = self._peek(block, timeout, items)
+
+            with self._file_lock:
+                self._set_queue_top(queue_top)
+
+                if isinstance(data, list):
+                    if len(data) > 0:
+                        self._update_length(self._length - len(data))
+                elif data is not None:
+                    self._update_length(self._length - 1)
+
+                self._get_event.set()
+                _LOGGER.debug("Returning data from get")
+                return data
+
+    def get_nowait(self):
+        """
+        Provides compatibility with stdlib Queue objects.
+
+        Equivalent to get(False).
+
+        Two methods are offered to support tracking whether enqueued tasks
+        have been fully processed by daemon consumer threads.
+        """
+        return self.get(block=False)
+
+    def task_done(self, items=1):
+        """
+        Provides compatibility with stdlib Queue objects.
+        Taken from the Python 3.5 stdlib: lib/python3.5/queue.py
+
+        Indicate that a formerly enqueued task is complete. Used by queue
+        consumer threads. For each get() used to fetch a task, a subsequent
+        call to task_done() tells the queue that the processing on the task is
+        complete.
+
+        If a join() is currently blocking, it will resume when all items have
+        been processed (meaning that a task_done() call was received for every
+        item that had been put() into the queue).
+
+        Raises a ValueError if called more times than there were items placed in the queue.
+        """
+        with self._all_tasks_done:
+            unfinished = self._unfinished_tasks - items
+            if unfinished <= 0:
+                if unfinished < 0:
```

Comment:
It's a style thing, but I don't like the nested checks on `unfinished`. How about something like:

```python
if unfinished < 0:
    raise ValueError('task_done() called too many times')
elif unfinished == 0:
    self._all_tasks_done.notify_all()
```